_posts/2023-07-14-types-and-programming-languages.markdown

```diff
@@ -0,0 +1,610 @@
+
+This we can enter in Dafny too:
+
+{% highlight javascript %}
+ghost function S(i: nat): iset<Term> {
+  if i == 0 then
+    iset{}
+  else
+    iset{True, False, Zero}
+    + (iset t1 <- S(i-1) :: Succ(t1))
+    + (iset t1 <- S(i-1) :: Pred(t1))
+    + (iset t1 <- S(i-1) :: IsZero(t1))
+    + (iset t1 <- S(i-1), t2 <- S(i-1), t3 <- S(i-1) :: If(t1, t2, t3))
+}
+ghost const AllTermsConstructively: iset<Term> := iset i: nat, t <- S(i) :: t
+{% endhighlight %}
+
+But now, we are left with the existential question: are these two sets the same?
+We rush in Dafny and write a lemma ensuring `AllTermsConstructively == AllTermsInductively` by invoking the lemma `InductiveAxioms()`, but... Dafny can't prove it.
+
+If you think deeply about it, how do you know that the two are the same? It seems obvious but why? It seems straightforward to prove that `AllTermsInductively <= AllTermsConstructively` because by definition, `AllTermsConstructively` obeys induction rules. But is it the smallest of such sets? But what if there was an element of `AllTermsConstructively` that is not in `AllTermsInductively`? It could actually happen if, instead of a datatype, we only had a trait, and some external user could implement new terms yet unknown to us.
+
+Here is Benjamin Pierce's proof sketch, then translated and verified in Dafny.
+
+1. First, prove `AllTermsInductively <= AllTermsConstructively` by showing that `AllTermsConstructively` satisfies the predicate `InductionCriteria`.
+2. Second, for any set `someset` satisfying the induction criteria, for every `i`, we prove by induction that *every set of terms `S(i)` is inside `someset`*.
+3. `AllTermsConstructively` being the *union* of all these `S(i)`, it is also contained
+in any set satisfying the induction criteria, including `AllTermsInductively` which is the smallest one, so `AllTermsConstructively <= AllTermsInductively`
+4. From 1. and 3. we obtain `AllTermsConstructively == AllTermsInductively`.
+
+Let's prove it in Dafny!
+
+## 0. Intermediate sets are cumulative
+
+First, we want to show that, for every `i <= j`, we have `S(i) <= S(j)` (set inclusion). We do this in two steps: first, we show this cumulative effect between two consecutive sets, and then
+between any two sets.
+
+We use the annotation `{:vcs_split_on_every_assert}` which makes Dafny verify each assertion independently, which, in this example, helps the verifier. Yes, [helping the verifier](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-verification-debugging-slow) is something we must occasionally do in Dafny.
+To futher control the situation, we use the annotation `{:induction false}` to ensure Dafny does not try to prove induction hypotheses by itself, which gives us control over the proof. Otherwise, Dafny can both automate the proof a lot (which is great!) and sometimes time out because automation is stuck (which is less great!). I left assertions in the code so that not only Dafny, but you too can understand the proof.
```

Comment:
```suggestion
To further control the situation, we use the annotation `{:induction false}` to ensure Dafny does not try to prove induction hypotheses by itself, which gives us control over the proof. Otherwise, Dafny can both automate the proof a lot (which is great!) and sometimes time out because automation is stuck (which is less great!). I left assertions in the code so that not only Dafny, but you too can understand the proof.
```